feat(step2): Introduce Calculator Class

leejuhanKr opened the pull request:
![image](https://github.com/wdj4606/kotlin-racingcar/assets/96106122/19386ac2-0a93-419f-ad25-04c397926869)

leejuhanKr committed:
feat: Introduce Operator enum for arithmetic operations

leejuhanKr committed:
feat: Introduce InputValidator for string validation

leejuhanKr committed:
feat: Implement Calculator

wdj4606 commented on src/test/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,35 @@
+            }
+    }
+    private fun parseOperator(token: String?): Operator {
+        return requireNotNull(token).let {
```
현재 token에 대해서 enum에서의 null체크와 calculator class에서의 null 체크가 동시에 일어나고 있어요!
enum에서의 null체크를 제거하던지, 해당 부분에서의 null체크를 제거하는게 좋을거 같습니다!

wdj4606 commented on src/test/kotlin/step2/InputValidator.kt:
```diff
@@ -0,0 +1,8 @@
+package step2
+
+object InputValidator {
+    fun requireNotNullOrBlank(value: String?, message: String = "Input must not be null or blank."): String {
```
위에 작성한 피드백으로 수정하신다면 해당 클래스는 필요 없어지겠져? 😄

wdj4606 commented on src/test/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,24 @@
+package step2
+
+enum class Operator(
```
와...enum class까지 아주 좋습니다! 👍 
typealias를 이용하여 해당 람다식을 추상화 해주셨는데, 이 부분을 그냥 인라인으로 명시하는게 오히려
명확하고 좋습니다!

wdj4606 reviewed (changes requested):
주한님! 생각보다 너무 잘 구현해주셔서 ㅋㅋㅋㅋ 본의 아니게 제가 욕심을 좀 냈습니다. 😄 
몇가지 피드백 남겼으니 피드백 확인 부탁드립니다!

leejuhanKr committed:
refactor: Change parameter type to nullable String

Modify the 'calc' function to accept a nullable String as input

leejuhanKr committed:
refactor: Improve readability by changing split delimiter in split expression

Change the split delimiter in the split expression from "\\s+".toRegex() to " ".toRegex() for improved readability. This modification enhances code clarity and simplifies the handling of whitespace separation.

leejuhanKr committed:
refactor: Simplify drop operation in the return statement

Simplify the drop operation in the return statement by removing the unnecessary use of asSequence(). This change is made to improve performance, especially in scenarios with smaller datasets where the asSequence() operation might introduce additional overhead.

leejuhanKr committed:
아 regex 제거해야되는데 모르고 안했음

leejuhanKr committed:
refactor: Simplify null-checks and enhance expression handling in Calculator

Remove unnecessary null-checks and simplify expression handling in the Calculator class. Utilize Kotlin's require function for cleaner expression validation, eliminating the need for explicit null or blank checks. The parseOperator and parseOperand functions are modified to receive non-nullable arguments, assuming the input has been validated at a higher level. This refactoring aims to improve code readability and maintainability.

leejuhanKr committed:
refactor: Remove OperatorFunc typealias in favor of inline lambda definition

Remove the OperatorFunc typealias and directly inline the lambda definition (Double, Double) -> Double where needed. This change is made for clarity and simplicity in expressing the lambda function without introducing a separate typealias. The modification aims to enhance code readability and reduce unnecessary abstractions.

wdj4606 commented on src/test/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    private fun parseOperator(token: String): Operator {
+        return token.let {
```
let 함수는 기본적으로 null인 인자들에 한해서 보통 사용을 많이합니다!
null인 아닌 타입에 대해선 인라인으로 바로 사용해주시는게 좋습니다!
`Operator.symbolOf(token)`

wdj4606 commented on src/test/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,34 @@
+        }
+    }
+
+    private fun parseOperand(token: String): Double {
```
음? 이부분 피드백 반영이 안된거 같습니다 ㅠ
계산에서의 Double과 float처럼 소수형은 정확한 계산이 되지 않습니다.
정수값에 대한 계산은 Int나 Long형을 사용해야만 정확한 계산이 가능합니다!

wdj4606 reviewed (approved):
주한님 빠른 피드백 반영 해주셨네요! 
머지할테니 피드백 확인해주시고 다음단계 진행해주세요 😄
